### ISSUE TITLE
XGBoost prediction not matching when model built on python and used in c++

### ISSUE THREAD

**Issue by harishvk27** [dmlc/xgboost#3423]
Bug in XGBoost C++ or Python library????.

Using python xgboost module i have built a multiclass classification model. The prediction is good when done on python, where as if i save this model file and use libxgboost.so c++ library to load model and then predict in c++, the predictions are not-correct.

Source of xgboost c++ library:https://github.com/dmlc/xgboost

I have pasted data-set used for building model. Python code for building model and C++ code where we load model and verify prediction for reference.

Not sure what's happening.

   cat xgBoost_MultiClass_DataSet.csv
    Feature-1,Feature-2,Feature-3,Classification-Type
    121,181500,18,0
    175,262500,26,0
    197,295500,29,0
    183,274500,27,0
    106,159000,15,0
    169,253500,25,0
    127,190500,19,0
    190,285000,28,0
    102,153000,15,0
    150,225000,22,0
    1324,1986000,198,1
    1667,2500500,250,1
    1393,2089500,208,1
    1269,1903500,190,1
    1369,2053500,205,1
    1285,1927500,192,1
    1308,1962000,196,1
    1085,1627500,162,1
    1855,2782500,278,1
    1536,2304000,230,1
    15437,23155500,2315,2
    11817,17725500,1772,2
    12270,18405000,1840,2
    14581,21871500,2187,2
    11515,17272500,1727,2
    14503,21754500,2175,2
    15225,22837500,2283,2
    13400,20100000,2010,2
    11399,17098500,1709,2
    14428,21642000,2164,2

Following is my python code to generate the model.

    import pandas as pd
    import sklearn
    import numpy as np
    import sys
    from sklearn import preprocessing
    
    
    df = pd.read_csv('MultiClass_DataSet.csv')
    
    #print(df.head())
    #print(df.shape)
    
    # first 3-columns are attributes
    # fifth column is classification-type
    data_x = df[df.columns[0:3]]
    data_y = df[df.columns[-1]]
    
    from sklearn.model_selection import train_test_split
    train_x, test_x, train_y, test_y = train_test_split(data_x, data_y, test_size=0.33, random_state=42)
    
    
    import xgboost as xgb
    xg_train = xgb.DMatrix(train_x, label=train_y)
    
    xg_test = xgb.DMatrix(test_x, label=test_y)
    
    param = {}
    # use softmax multi-class classification
    param['objective'] = 'multi:softmax'
    # scale weight of positive examples
    param['eta'] = 0.1
    param['silent'] = 1
    param['nthread'] = 1
    param['num_class'] = 3
    param['max_depth'] = 10
    param['colsample_bytree'] = 0.7
    param['alpha'] = 0.03
    param['subsample'] = 0.8
    
    watchlist = [(xg_train, 'train'), (xg_test, 'test')]
    
    bst = xgb.train(param, xg_train, 300, watchlist, early_stopping_rounds=50)
    
    pred =  bst.predict(xg_test)
    
    from sklearn.metrics import accuracy_score
    predictions = [round(value) for value in pred]
    accuracy = accuracy_score(test_y, predictions)
    print("Accuracy: %.2f%%" % (accuracy * 100.0))
    
    
    bst.save_model('xgBoost_MultiClass_Classification.model')

Following is my C++ XGBoost model used to verify model.

    #include <stdio.h>
    #include <iostream>
    #include <vector>
    #include <memory>
    #include <stdexcept>
    #include <xgboost/c_api.h>
    #include <assert.h>
    
    typedef void *BoosterHandle;
    typedef void *DMatrixHandle;
    typedef uint64_t bst_ulong;
    
    //! Pointer to the xgboost model
    BoosterHandle g_boosterModel;
    
    //! Number of rows in the input. This will be passed to 
    //! XGDMatrixCreateFromMat function to create the DMatrix
    const DMatrixHandle *g_trainHandle = nullptr;
    
    //! Length of the train handle
    bst_ulong g_trainHandleLength = 0;
    
    bst_ulong g_testInputRowCount = 1;
    
    //! Filler for missing values
    const float g_missing = -999.9f;
    
    
    void CreateModel(const std::string& configFile)
    {
        int result=-1;
    
        result = XGBoosterCreate(
                                 g_trainHandle,
                                 g_trainHandleLength,
                                 &g_boosterModel
                                 );
        if(result == -1 || g_boosterModel == nullptr)
        {
           printf("failed1");
        }
    
        result = XGBoosterLoadModel(g_boosterModel,configFile.c_str());
    
        if(result == -1)
        {
           printf("failed2");
        }
    }
    
    std::vector<double> Evaluate(const std::vector<double>& features)
    {
        DMatrixHandle inputDMatrixHandle;
        std::vector<double> emptyVector;
    
        // Make sure the size of the input is only one
        if(g_testInputRowCount != 1)
        {
            printf("failed:2:1"); 
            return emptyVector; 
        }
    
        // Convert inputFeatures to DMatrix format, so that we can pass it to 
        // The predict function of xgboost library
        int result = XGDMatrixCreateFromMat(
                                        (float *) &features[0],   // The input data
                                        g_testInputRowCount,           // Number of rows in the input 
                                        features.size() ,              // Number of columns in the input
                                        g_missing,                     // Filler for missing values 
                                        &inputDMatrixHandle            // Handle to DMatrix of input
                                       );
    
        // Was not able to create the DMatrix
        if(result == -1)
        {
            printf("failed:2:2");
            return emptyVector;
        }
    
        const float *output;
        bst_ulong outputLength;
    
        // Do the prediction on the input features
        result = XGBoosterPredict(g_boosterModel,            // Handle to xgboost model
                         inputDMatrixHandle,                 // DMatrix of input 
                         0,                                  // Option mask 0: normal prediction 
                         0,                                  // Number of trees used for prediction
                         &outputLength,                      // Length of the ouput 
                         &output                             // actual output
                         );
    
        // Model has failed to predict due to some reason
        if(result == -1)
        {
            printf("failed:2:3");
            return emptyVector;
        }
    
        // Convert the output from xgboost predict to vector
        std::vector<double> outputVector(output, output + outputLength);
        result = XGDMatrixFree(inputDMatrixHandle);
    
        if(result == -1)
        {
            printf("failed:2:4");
            return emptyVector;
        }
    
        return outputVector;
    }
    
    void TestModel()
    {
        CreateModel("/tmp/xgBoost_MultiClass_Classification.model");
    
        std::vector<double> features {121.0,181500.0,18.0};
        
        std::vector<double> res = Evaluate(features);
        std::cout << "Prediction:" << res[0] << std::endl;
    
        std::vector<double> features2{1324.0,1986000.0,198.0};
        std::vector<double> res2 = Evaluate(features2);
        std::cout << "Prediction:" << res2[0] << std::endl;
    
        std::vector<double> features3{15437.0,23155500.0,2315.0,2.0};
        std::vector<double> res3 = Evaluate(features3);
        std::cout << "Prediction:" << res3[0] << std::endl;
    }
    
    int main(int argc, char **argv)
    {
       try {
           TestModel();
       }
       catch(...){
          std::cout << "exception occured" << std::endl;
       }
    
    }
    

Expected predicted Classification-Type values should be 0,1 and 2.
Where as C++ loaded model is predicting 0,0,0.

What i have tried:
(1) Verified prediction on python
(2) Downloaded latest libxgboost library source and build library freshly
(3) There were quite a few similar issues raised earlier but most of them where because of not handling category-type feature attributes, if they are not converted properly the then xgboost could miss-interpret. So got to a data-set which has no invalid column values, simple to look at and predictable.

Still seeing this issue..

**Comment by hcho3** [comment]
You should use `float` for all real numbers, not `double`. So use `std::vector<float>& features` instead of `std::vector<double>& features`.

**Comment by harishvk27** [comment]
Thanks.. this one worked..

I didn't understand why there wasn't any exception thrown if something was wrong.
